### ISSUE TITLE
Import App from Catalog

### ISSUE THREAD

**Issue by dkjones9602** [rzander/sccmclictr#106]
Will the Import App from Catalog stop working in the next SCCM upgrade when they deprecate the catalog?

**Comment by rzander** [comment]
Good question, I've never tried it without the App Catalog Role... Will check it...

**Comment by rzander** [comment]
I can confirm that it will **NOT** work...

**Comment by dkjones9602** [comment]
Can i assume that feature will of client center will go away?
Unfortunately that is highly used here.

**Comment by rzander** [comment]
I have not found any documentations how this AppCatalog API looks like over the MP... If someone has some more details please let me know... otherwise this feature will go away....

**Comment by Xento** [comment]
Maybe this isn't needed anymore with SCCM 1906:
https://docs.microsoft.com/en-us/sccm/apps/deploy-use/install-app-for-device

Am Do., 8. Aug. 2019 um 13:32 Uhr schrieb dkjones9602 <
notifications@github.com>:

> Can i assume that feature will of client center will go away?
> Unfortunately that is highly used here.
>

**Comment by dkjones9602** [comment]
That feature in the console doesn't help the tech in the field. Roger has
built an amazing all in one tool that doesn't crash frequently like the
console and can be put on a thumb drive.

**Comment by Xento** [comment]
The console is very stable for us.
Sometimes I have it open for many days.

Maybe something can be implemented that calls this function from this tool
via WMI?

Am Do., 8. Aug. 2019 um 14:20 Uhr schrieb dkjones9602 <
notifications@github.com>:

**Comment by dkjones9602** [comment]
We upgraded to 1906 and the Import App feature is still working. I wonder
if the components would be there on fresh install instead of an upgrade?  I
don't get a complete list. It always truncates at the apps starting with
'U'. Do you have cap on the number of returns? If so, can it be overridden?

**Comment by theaquamarine** [comment]
No, the app catalog role can't be added on 1906, but if you've already got one it won't be removed. It's not been required since 1806 but was supported, so if you already had one it'd work but not be used by clients. 

1902 is the most recent baseline release, so a new install would have chance to install it before upgrading, if you really wanted to...

https://docs.microsoft.com/en-us/sccm/apps/plan-design/plan-for-and-configure-application-management#bkmk_remove-appcat

**Comment by NathanHopkins6283** [comment]
Hi Roger.  I just upgrade our infrastructure  to 1906 and we miss this functionality.  From what I've been able to piece together after the catalog roles are removed, this info is available via WMI by querying the Primary.

**Get all apps that require approval:**
$primarySCCMServer = '<PrimaryServer>'
$namespace = "ROOT\SMS\site_<SiteCode>"
$classname = "SMS_ApplicationAssignment"

$apps = Get-WmiObject -Class $classname -ComputerName $primarySCCMServer -Namespace $namespace | Where-Object {$_.RequireApproval -eq 'True'} | Select ApplicationName, AssignedCI_UniqueID

$masterArray = @()
ForEach ($item in $apps){
    $array = new-object PSObject
    $array | add-member -type NoteProperty -Name ApplicationName -Value $item.ApplicationName
    $array | add-member -type NoteProperty -Name AssignedCI_UniqueID -Value ($item.AssignedCI_UniqueID).Substring(0, $item.AssignedCI_UniqueID.LastIndexOf('/'))
    $masterArray += $array
}

**Next you'll need to select the application and corresponding AssignedCI_UniqueID and invoke a WMI method against the Primary:**
$primarySCCMServer = '<PrimaryServer>'
$sccmClient = '<sccmClient>'
$AutoInstall = $true
$comments = "Test Comment"
$namespace = "ROOT\SMS\site_<SiteCode>"
$machine = Get-WmiObject -ComputerName $primarySCCMServer -Namespace $namespace -Query "SELECT * FROM SMS_R_SYSTEM WHERE Name = $sccmClient"
$clientGuid = $machine.SMSUniqueIdentifier
Invoke-WmiMethod -ComputerName $primarySCCMServer -Path "SMS_ApplicationRequest" -Namespace $NameSpace -Name CreateApprovedRequest -ArgumentList @($AppID, $AutoInstall, $clientGuid, $comments)

Note: $AppID in this example is the AssignedCI_UniqueID from the $masterArray

I hope this helps and gets into the next revision.

This assumes this guide has been followed: https://www.petervanderwoude.nl/post/real-time-application-installation-for-devices/

**Comment by rzander** [comment]
Application (pre) Approval is a cool thing if you have a user loggedon on the target device. But Client Center seems to be the wrong tool for that, as all to logic has to run on the Site Server... using the AdminService (REAT API) could be an option, have to check that....

**Comment by NathanHopkins6283** [comment]
The WMI query to get the pre-approval apps takes a while to complete.  I
have a feeling that the AdminService will be a lot faster.  I've been
having trouble getting the service working in my environment.  I have a
feeling it's a certificate configuration issue.  I didn't know you needed a
user logged on to the target device.  I'll have to test this out - our
deployments are set to run even without a user logged on.

Thanks for getting back to me by the way.

Nate

**Comment by NathanHopkins6283** [comment]
I just tested this out on a device where no user was logged on and it
installed the app.  If the app deployment  is configured whether or not a
user is logged on seems to be the key.
